parla/task_runtime.py

```diff
@@ -973,45 +958,76 @@ def shuffled(lst: Iterable[_T]) -> List[_T]:
     return lst
 
 class Scheduler(ControllableThread, SchedulerContext):
+    # See __init__ function below for comments on the functionality of these members
     _environments: TaskEnvironmentRegistry
-    # Worker threads for each device.
     _worker_threads: List[WorkerThread]
+    _free_worker_threads: Deque[WorkerThread]
     _available_resources: ResourcePool
-    period: float
-    max_worker_queue_depth: int
+    period: float # TODO: Figure out what this is for
 
     def __init__(self, environments: Collection[TaskEnvironment], n_threads: int = None, period: float = 0.01,
                  max_worker_queue_depth: int = 2):
+        # ControllableThread: __init__ sets it to run
+        # SchedulerContext: No __init__
         super().__init__()
+
         # TODO(lhc): for now, assume that n_threads is always None.
         #            Each device needs a dedicated thread.
         n_threads = sum(d.resources.get("vcus", 1) for e in environments for d in e.placement)
+
+        # TODO: Figure out what these are for
         self._environments = TaskEnvironmentRegistry(*environments)
+
+        # Empty list for storing reported exceptions at runtime
         self._exceptions = []
-        self._active_task_count = 1 # Start with one count that is removed when the scheduler is "exited"
+
+        # Start with one count that is removed when the scheduler is "exited"
+        self._active_task_count = 1
+
+        # TODO: Delete this
         self.max_worker_queue_depth = max_worker_queue_depth
+
+        # TODO: Figure out what this is for
         self.period = period
+
+        # TODO: Figure out what this is for
```

Comment:
Why is it questionable? I think this is necessary. For example, any thread can spawn task and it causes race condition.